src/Orleans/Lifecycle/LifecycleObservable.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans
+{
+    public class LifecycleObservable<TStage> : ILifecycleObservable<TStage>, ILifecycleObserver
+    {
+        private readonly ConcurrentDictionary<object, OrderedObserver> subscribers;
+        private readonly Logger logger;
+        private TStage highStage;
+
+        public LifecycleObservable(Logger logger)
+        {
+            this.logger = logger?.GetLogger(GetType().Name);
+            subscribers = new ConcurrentDictionary<object, OrderedObserver>();
+        }
+
+        public async Task OnStart()
+        {
+            try
+            {
+                foreach (IGrouping<TStage, OrderedObserver> observerGroup in subscribers.Values
+                    .GroupBy(orderedObserver => orderedObserver.Stage)
+                    .OrderBy(group => group.Key))
+                {
+                    highStage = observerGroup.Key;
+                    await Task.WhenAll(observerGroup.Select(orderedObserver => WrapExecution(orderedObserver.Observer.OnStart)));
```

Comment:
In your suggested whenall line, what happens if there are 10 observers in the group and the third one does this?

Task OnStop()
{
    throw new NotImplementedException():
}

The remaining 7 OnStop calls never get called, correct?  This is what the wrapper is for.  With the wrapper all OnStop calls are executed, regardless of errors.